velox/functions/sparksql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,136 @@
+ * limitations under the License.
+ */
+
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/sparksql/tests/SparkFunctionBaseTest.h"
+
+namespace facebook::velox::functions::sparksql::test {
+namespace {
+
+using namespace facebook::velox::test;
+
+class ArrayShuffleTest : public SparkFunctionBaseTest {
+ protected:
+  template <typename T>
+  VectorPtr
+  testShuffle(const VectorPtr& input, int64_t seed, int32_t partitionId = 0) {
+    setSparkPartitionId(partitionId);
+    return evaluate(
+        fmt::format("shuffle(c0, {})", seed), makeRowVector({input}));
+  }
+
+  template <typename T>
+  void compareResult(
+      const VectorPtr& result,
+      const VectorPtr& expected,
+      bool equal) {
+    DecodedVector decodedResult(*result.get());
+    exec::VectorReader<Array<T>> readerResult(&decodedResult);
+    DecodedVector decodedExpected(*expected.get());
+    exec::VectorReader<Array<T>> readerExpected(&decodedExpected);
+    for (auto i = 0; i < decodedExpected.size(); i++) {
+      auto resultArray = readerResult[i].materialize();
+      auto expectedArray = readerExpected[i].materialize();
+      if (equal) {
+        ASSERT_TRUE(std::equal(
+            resultArray.begin(), resultArray.end(), expectedArray.begin()));
+      } else {
+        if (resultArray.size() > 2) {
+          ASSERT_FALSE(std::equal(
+              resultArray.begin(), resultArray.end(), expectedArray.begin()));
+        }
+        ASSERT_TRUE(std::is_permutation(
+            resultArray.begin(), resultArray.end(), expectedArray.begin()));
+      }
+    }
+  }
+};
+
+TEST_F(ArrayShuffleTest, basic) {
+  auto input = makeArrayVector<int64_t>({{1, 2, 3, 4, 5}});
+  auto result = makeArrayVector<int64_t>({{3, 5, 4, 1, 2}});
+  auto stringInput = makeArrayVector<std::string>({{"a", "b", "c", "d"}});
+  auto stringResult = makeArrayVector<std::string>({{"a", "c", "b", "d"}});
+  compareResult<int64_t>(testShuffle<int64_t>(input, 0), result, true);
+  compareResult<std::string>(
+      testShuffle<std::string>(stringInput, 0), stringResult, true);
+
+  // Assert results are different with different seeds / partition ids.
+  compareResult<int64_t>(
+      testShuffle<int64_t>(input, 0, 0),
+      testShuffle<int64_t>(input, 0, 1),
+      false);
+  compareResult<int64_t>(
+      testShuffle<int64_t>(input, 1, 0),
+      testShuffle<int64_t>(input, 0, 0),
+      false);
+}
+
+TEST_F(ArrayShuffleTest, nestedArrays) {
+  using innerArrayType = std::vector<std::optional<int64_t>>;
+  using outerArrayType =
+      std::vector<std::optional<std::vector<std::optional<int64_t>>>>;
+  innerArrayType a{1, 2, 3, 4};
+  innerArrayType b{5, 6};
+  innerArrayType c{6, 7, 8};
+  outerArrayType row1{{a}, {b}};
+  outerArrayType row2{std::nullopt, std::nullopt, {a}, {b}, {c}};
+  outerArrayType row3{{}};
+  outerArrayType row4{{{std::nullopt}}};
+  auto input =
+      makeNullableNestedArrayVector<int64_t>({{row1}, {row2}, {row3}, {row4}});
```

Comment:
same, not nullable

Reply:
Changed to `makeNestedArrayVectorFromJson`, thanks.